File: Resoluções-das-Listas/Lista-1/Lista-1-Problema-4.cpp
#include <iostream>
#include <cmath>

using namespace std;

// Função a ser integrada
double f(double x)
{
    return sin(x) / (x + 0.1);
}

// Método de Simpson para integrar a função f de a até b com N subintervalos
double simpson(double a, double b, int N)
{
    double h = (b - a) / N;
    double sum = f(a) + f(b);

    for (int i = 1; i < N; i++)
    {
        double x = a + i * h;
        if (i % 2 == 0)
        {
            sum += 2 * f(x);
        }
        else
        {
            sum += 4 * f(x);
        }
    }

    return (h / 3) * sum;
}

int main()
{
    double a, b;
    int N;

    cout << "Digite o limite inferior de integração (a): ";
    cin >> a;
    cout << "Digite o limite superior de integração (b): ";
    cin >> b;
    cout << "Digite o número de subintervalos (N): ";
    cin >> N;

    double result = simpson(a, b, N);
    cout << "O valor da integral é: " << result << endl;

    return 0;
}

File: Resoluções-das-Listas/Lista-1/Lista-1-Problema-4.py
import math

# Função a ser integrada


def f(x):
    return math.sin(x) / (x + 0.1)

# Método de Simpson para integrar a função f de a até b com N subintervalos


def simpson(a, b, N):
    h = (b - a) / N
    sum = f(a) + f(b)

    for i in range(1, N):
        x = a + i * h
        if i % 2 == 0:
            sum += 2 * f(x)
        else:
            sum += 4 * f(x)

    return (h / 3) * sum


# Entrada do usuário
a = float(input("Digite o limite inferior de integração (a): "))
b = float(input("Digite o limite superior de integração (b): "))
N = int(input("Digite o número de subintervalos (N): "))

result = simpson(a, b, N)
print(f"O valor da integral é: {result}")


def find_stable_N(a, b, tolerance=1e-6):
    N = 10
    previous_result = simpson(a, b, N)
    while True:
        N *= 2
        current_result = simpson(a, b, N)
        if abs(current_result - previous_result) < tolerance:
            break
        previous_result = current_result
    return N, current_result


a = float(input("Digite o limite inferior de integração (a): "))
b = float(input("Digite o limite superior de integração (b): "))

N, result = find_stable_N(a, b)
print(f"O valor de N requerido é: {N}")
print(f"O valor da integral é: {result}")
